Add mean and peak frequency to SAT features output, add parameter "center" option to pitch

eaogorman opened the pull request.

NickleDave commented:
Hey @eaogorman just checking in about this -- did you mean to close this pull request?

Looks like you were adding some features from avn?
We're definitely open to contributions!

Let me know what you're thinking, we can talk through it here or jump on a quick Zoom call if that's easier

eaogorman commented:
Hi @NickleDave! Yes, sort of, a bit of an accidental/premature PR, oops!

I include mean and peak frequency in my analysis pipeline. I've recently integrated vocalpy, forked, added a function for peak frequency in [src/vocalpy/feature/_sat.py](https://github.com/vocalpy/vocalpy/pull/209/commits/430380dfd89246202a3fa402da32eee0c2229b2c#diff-b93905ce9c14cfae09857503956f4aa9cb876b66ff26adfd987bc22606f96499)], after line [143](https://github.com/vocalpy/vocalpy/blob/2d9ae6a9c532938911edbd2f8b4f431af6792295/src/vocalpy/feature/_sat.py#L143) sort of adapted from avn, but also just bounding frequencies between min_freq, max_freq, finding frequencies with highest power.

I added mean and peak frequency to Features output after line [627](https://github.com/vocalpy/vocalpy/blob/2d9ae6a9c532938911edbd2f8b4f431af6792295/src/vocalpy/feature/_sat.py#L627). These features could be included as an option for other users, as input to the sat function, maybe as a list of feature names with the default set as without mean and peak frequency?

Also, when I compute pitch using [librosa.yin](https://librosa.org/doc/0.10.2/generated/librosa.yin.html#librosa.yin), I sometimes play around with the center parameter. When center=True, the left/right edges are padded with zeros when pad_mode="constant" and my estimate for mean pitch of the segment is skewed by the first and last values. Sometimes I include center=True, but pad_mode="empty" to maintain proper length of output (matching the other features). I tried including this, but I think it threw an error. Maybe something to consider adding as a parameter for the pitch function.

eaogorman commented on src/vocalpy/feature/_sat.py:
```diff
@@ -559,7 +623,9 @@ def sat(
         {
             "amplitude": (["channel", "time"], amp_),
             "pitch": (["channel", "time"], pitch_),
+            "pitch_no_center": (["channel", "time"], pitch_(center=False))
```
If center=False, then output is two values smaller than other output. Would work if center=True and pad_mode="empty", I think, same behavior as center=False, but maintain shape. pitch_[1:-1] == pitch_(center=False), and I think pitch_[1:-1] == pitch_(center=True, pad_mode="empty")[1:-1] and pitch_(center=True, pad_mode="empty")[0] and pitch_(center=True, pad_mode="empty")[-1] would be undefined. Need to test, tbd.

eaogorman commented on src/vocalpy/feature/_sat.py:
```diff
@@ -141,6 +141,64 @@ def mean_frequency(
+    min_freq: float = 380.0,
+    max_freq: float = 11025.0,
+) -> npt.NDArray:
+    """Calculate peak frequency.
```
Need to clean up docstring.
